tenders/spt/spt_module_net.c

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright (c) 2015-2018 Contributors as noted in the AUTHORS file
+ *
+ * This file is part of Solo5, a sandboxed execution environment.
+ *
+ * Permission to use, copy, modify, and/or distribute this software
+ * for any purpose with or without fee is hereby granted, provided
+ * that the above copyright notice and this permission notice appear
+ * in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
+ * WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
+ * AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR
+ * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
+ * OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
+ * NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+/*
+ * spt_module_net.c: Network device module.
+ */
+
+#include <assert.h>
+#include <err.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <ifaddrs.h>
+#include <inttypes.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+#include <seccomp.h>
+
+#include "spt.h"
+
+#if defined(__linux__)
+
+/*
+ * Linux TAP device specific.
+ */
+#include <sys/socket.h>
+#include <linux/if.h>
+#include <linux/if_tun.h>
+
+#elif defined(__FreeBSD__)
+
+#include <net/if.h>
+
+#elif defined(__OpenBSD__)
+
+#include <sys/socket.h>
+#include <net/if.h>
+
+#else /* !__linux__ && !__FreeBSD__ && !__OpenBSD__ */
+
+#error Unsupported target
+
+#endif
+
+static char *netiface;
+static int netfd = -1;
+static uint8_t guest_mac[6];
+static int cmdline_mac = 0;
+
+/*
+ * Attach to an existing TAP interface named 'ifname'.
+ *
+ * Returns -1 and an appropriate errno on failure (ENOENT if the interface does
+ * not exist), and the tap device file descriptor on success.
+ */
+static int tap_attach(const char *ifname)
+{
+    int fd;
+
+    /*
+     * Syntax @<number> indicates a pre-existing open fd, so just pass it
+     * through without any checks.
+     */
+    if (ifname[0] == '@') {
+        fd = atoi(&ifname[1]);
```

Comment:
does this require error checking?

Reply:
I believe the `fcntl` below will fail if you pass it something absurd, but it might make sense to verify that `fd >`  `0` or `1` or `2`?

Reply:
I was thinking of cases like `@12m` silently parsed as `12`, but in any case this shouldn't be a  security issue.

Reply:
Agreed, I guess the belt-and-suspenders solution would be to use `strtol` to ensure the whole string was consumed, then checking for fd collisions with existing FDs, but I appreciate there's a complexity trade-off that I don't have a strong opinion about.

Reply:
This is an "infrastructure" option intended for higher-level stacks/automation, there's not much point/no straightforward way to check at this level if the user has passed in an _incorrect_ `fd`.

OTOH, if an _invalid_ number is passed in, you will effectively get `fd=0` with the current code since there's no way to distinguish an error with `atoi` and `0` may be valid if whatever exec'd solo5-spt closed STDIN and then immediately opened a TAP device.

If you strongly care about the above case in particular please file a separate issue for it since the same code is also in `hvt` and thus best fixed in one go once this is merged.

Reply:
One more thing: If we do decide to check for the "invalid number" case, it's probably easier to use `sscanf` with a `%u` format specifier rather than `strtoul`.

Reply:
Just a note from what I found on my twitter feed:

> A note about the ato{i,l,f} functions – their behavior is undefined if the input cannot be parsed correctly. These functions need not set errno even. Except for behavior on error, atoi is equivalent to (int)strtol(nptr, (char **)NULL, 10). The strto{d,l,ul} functions should always be preferred, because they provide proper error reporting and choice of base.

here: https://begriffs.com/posts/2019-01-19-inside-c-standard-lib.html